### PR TITLE
Fix `make help`

### PR DESCRIPTION
**Pull request by luksa**

### CODE REVIEW

**Review by jwendell — DISMISSED** [review]
The common file should not be edited directly. It comes from istio/common-files and is updated regularly. We should either adapt to it (maybe overriding some stuff) or, if there's a bug, fix it in upstream.

**Comment by luksa** [comment]
We need a way to improve/override the targets in the common file. The change I made to the `help` target doesn't make much sense in the upstream code base, because there are only a few targets and they aren't organized into topics. But in our operator, we have many targets and they're all nicely grouped. Before we switched to using the upstream Makefiles, this was the output of `make help`:
```
$ make help

Usage:
  make <target>

General
  help             Display this help.

Testing
  test             Run tests.

Build
  build            Build manager binary.
  run              Run a controller from your host.
  docker-build     Build docker image with the manager.
  docker-push      Push docker image with the manager.
  docker-buildx    Build and push docker image for the manager for cross-platform support

Deployment
  install          Install CRDs into the K8s cluster specified in ~/.kube/config.
  uninstall        Uninstall CRDs from the K8s cluster specified in ~/.kube/config. Call with ignore-not-found=true to ignore resource not found errors during deletion.
  deploy           Deploy controller to the K8s cluster specified in ~/.kube/config.
  deploy-yaml      Outputs YAML manifests needed to deploy the controller
  deploy-olm       Builds and pushes the operator OLM bundle and then deploys the operator using OLM
  undeploy         Undeploy controller from the K8s cluster specified in ~/.kube/config. Call with ignore-not-found=true to ignore resource not found errors during deletion.
  undeploy-olm     Undeploys the operator from the cluster (used only if operator was installed via OLM)
  deploy-example-openshift  Deploy an example Istio resource on OpenShift
  deploy-example-kubernetes  Deploy an example Istio resource on Kubernetes

Generated Code & Resources
  gen-manifests    Generate WebhookConfiguration, ClusterRole and CustomResourceDefinition objects.
  gen-code         Generate code containing DeepCopy, DeepCopyInto, and DeepCopyObject method implementations.
  gen-charts       Pull charts from maistra/istio repository
  gen-check        Verifies that changes in generated resources have been checked in
  update-istio     Updates the Istio commit hash to latest on ${ISTIO_BRANCH_30}
  patch-istio-images  Patch the Istio images in the ClusterServiceVersion with the right tags

Build Dependencies
  kustomize        Download kustomize locally if necessary. If wrong version is installed, it will be removed before downloading.
  controller-gen   Download controller-gen locally if necessary. If wrong version is installed, it will be overwritten.
  envtest          Download envtest-setup locally if necessary.
  bundle           Generate bundle manifests and metadata, then validate generated files.
  bundle-build     Build the bundle image.
  bundle-push      Push the bundle image.
  bundle-publish   Create a PR for publishing in OperatorHub
  opm              Download opm locally if necessary.
  catalog-build    Build a catalog image.
  catalog-push     Push a catalog image.

Linting
  lint-scripts     runs linters against all shell scripts
  lint-go          runs linters against Go code
  lint-yaml        runs linters against YAML files
  lint-helm        runs linters against helm charts
  lint-bundle      runs linters against OLM metadata bundle
  lint-watches     checks if the operator watches all resource kinds present in Helm charts
  lint             runs all linters
```

This is the output after the change:
```
$ make help
build                Build manager binary.
bundle-build         Build the bundle image.
bundle               Generate bundle manifests and metadata, then validate generated files.
bundle-publish       Create a PR for publishing in OperatorHub
bundle-push          Push the bundle image.
catalog-build        Build a catalog image.
catalog-push         Push a catalog image.
controller-gen       Download controller-gen locally if necessary. If wrong version is installed, it will be overwritten.
deploy-example-kubernetes Deploy an example Istio resource on Kubernetes
deploy-example-openshift Deploy an example Istio resource on OpenShift
deploy               Deploy controller to the K8s cluster specified in ~/.kube/config.
deploy-olm           Builds and pushes the operator OLM bundle and then deploys the operator using OLM
deploy-yaml          Outputs YAML manifests needed to deploy the controller
docker-build         Build docker image with the manager.
docker-buildx        Build and push docker image for the manager for cross-platform support
docker-push          Push docker image with the manager.
envtest              Download envtest-setup locally if necessary.
gen-charts           Pull charts from maistra/istio repository
gen-check            Verifies that changes in generated resources have been checked in
gen-code             Generate code containing DeepCopy, DeepCopyInto, and DeepCopyObject method implementations.
gen-manifests        Generate WebhookConfiguration, ClusterRole and CustomResourceDefinition objects.
help                 Show this help
install              Install CRDs into the K8s cluster specified in ~/.kube/config.
kustomize            Download kustomize locally if necessary. If wrong version is installed, it will be removed before downloading.
lint-bundle          runs linters against OLM metadata bundle
lint                 runs all linters
lint-watches         checks if the operator watches all resource kinds present in Helm charts
opm                  Download opm locally if necessary.
patch-istio-images   Patch the Istio images in the ClusterServiceVersion with the right tags
.PHONY               Generate nightly bundle
.PHONY               Build and push nightly docker image with the manager.
.PHONY               Generate everything
.PHONY               Runs the operator scorecard test. Needs a valid k8s cluster as pointed by the KUBECONFIG variable
run                  Run a controller from your host.
test                 Run tests.
undeploy-olm         Undeploys the operator from the cluster (used only if operator was installed via OLM)
undeploy             Undeploy controller from the K8s cluster specified in ~/.kube/config. Call with ignore-not-found=true to ignore resource not found errors during deletion.
uninstall            Uninstall CRDs from the K8s cluster specified in ~/.kube/config. Call with ignore-not-found=true to ignore resource not found errors during deletion.
update-istio         Updates the Istio commit hash to latest on ${ISTIO_BRANCH_30}
```

This is now pretty useless, since it's no longer organized by topic.

**Comment by jwendell** [comment]
If you think organizing by topics is better, what does prevent us to implement this enhancement in upstream?

**Comment by luksa** [comment]
Upstream doesn't have that many targets:
```
$ make help
benchtest            Runs all benchmarks
build                Builds all go binaries.
clean                Cleans all the intermediate files and folders previously generated.
docker               Build all docker images
docker.push          Build all docker images and push to
docker.save          Build docker images and save to tar.gz
format               Auto formats all code. This should be run before sending a PR.
help                 Show this help
lint                 Runs all linters.
push                 Build and push docker images to registry defined by $HUB and $TAG
test                 Runs all unit tests
```

Also, if we were to add topics, we'd need to change the order of a lot of targets in the Makefile.

**Comment by jwendell** [comment]
sgtm

File: common/Makefile.common.mk [hunk 1]
```diff
@@ -116,6 +116,6 @@ tidy-docker:
 
 # help works by looking over all Makefile includes matching `target: ## comment` regex and outputting them
 help: ## Show this help
-	@egrep -h '^[a-zA-Z_\.-]+:.*?## .*$$' $(MAKEFILE_LIST) | sort  | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-20s\033[0m %s\n", $$1, $$2}'
```
Comment (dgn) [inline common/Makefile.common.mk]: we can move this target to `Makefile.overrides.mk`, we just need to manually define MAKEFILE_LIST as well then:

```Makefile
MAKEFILE_LIST = Makefile Makefile.core.mk Makefile.overrides.mk

.PHONY: help
help:
	@awk 'BEGIN {FS = ":.*##"; printf "\nUsage:\n make \033[36m<target>\033[0m\n"} /^[a-zA-Z_0-9-]+:.*?##/ { printf " \033[36m%-15s\033[0m %s\n", $$1, $$2 } /^##@/ { printf "\n\033[1m%s\033[0m\n", substr($$0, 5) } ' $(MAKEFILE_LIST)
```

That way it works and we don't need to change `Makefile.common.mk` or `Makefile`

**Comment by dgn** [comment]
Let's get this merged! The current `make help` is really broken

**Comment by dgn** [comment]
```
21:47:06 p1 {make-help} ~/dev/istio-operator$ make help

Usage:
  make <target>

Testing
  test                       Run tests.
  test.scorecard             Run the operator scorecard test against an existing cluster.
  test.integration.ocp       Run the integration tests against an existing OCP cluster.
  test.integration.kind      Deploy a KinD cluster and run the integration tests against it.

Build
  build                      Build manager binary.
  run                        Run a controller from your host.
  docker-build               Build docker image.
  docker-push                Push docker image.
  docker-push-nightly        Build and push nightly docker image.
  docker-buildx              Build and push docker image with cross-platform support.

Deployment
  install                    Install CRDs into an existing cluster.
  uninstall                  Uninstall CRDs from an existing cluster. Call with ignore-not-found=true to ignore resource not found errors during deletion.
  deploy                     Deploy controller to an existing cluster.
  deploy-yaml                Output YAML manifests used by `deploy`.
  deploy-openshift           Deploy controller to an existing OCP cluster.
  deploy-yaml-openshift      Output YAML manifests used by `deploy-openshift`.
  deploy-olm                 Build and push the operator OLM bundle and deploy the operator using OLM.
  undeploy                   Undeploy controller from an existing cluster. Call with ignore-not-found=true to ignore resource not found errors during deletion.
  undeploy-olm               Undeploy the operator from an existing cluster (used only if operator was installed via OLM).
  deploy-example             Deploy an example Istio resource to an existing OCP cluster. Same as `deploy-example-openshift`.
  deploy-example-openshift   Deploy an example Istio resource to an existing OCP cluster.
  deploy-example-kubernetes  Deploy an example Istio resource on an ecisting cluster.

Generated Code & Resources
  gen-manifests              Generate WebhookConfiguration and CustomResourceDefinition objects.
  gen-api                    Generate API types from upstream files.
  gen-code                   Generate code containing DeepCopy, DeepCopyInto, and DeepCopyObject method implementations.
  gen-charts                 Pull charts from istio repository.
  gen                        Generate everything.
  gen-check                  Verify that changes in generated resources have been checked in.
  update-istio               Update the Istio commit hash in the 'latest' entry in versions.yaml to the latest commit in the branch.

Build Dependencies
  helm                       Download helm to bin directory. If wrong version is installed, it will be overwritten.
  controller-gen             Download controller-gen to bin directory. If wrong version is installed, it will be overwritten.
  envtest                    Download envtest-setup to bin directory.
  bundle                     Generate bundle manifests and metadata, then validate generated files.
  bundle-build               Build the bundle image.
  bundle-push                Push the bundle image.
  bundle-publish             Create a PR for publishing in OperatorHub.
  bundle-nightly             Generate nightly bundle.
  bundle-publish-nightly     Publish nightly bundle.
  patch-istio-crd            Update Istio CRD's openAPIV3Schema values.
  catalog-build              Build a catalog image.
  catalog-push               Push a catalog image.

Linting
  lint-bundle                Run linters against OLM metadata bundle.
  lint-watches               Checks if the operator watches all resource kinds present in Helm charts.
  lint                       Run all linters.
  format                     Auto-format all code. This should be run before sending a PR.
  help                       Show this help.
```

File: Makefile.core.mk [hunk 2]
```diff
@@ -148,16 +148,16 @@ export
 test: envtest ## Run tests.
 	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) --bin-dir $(LOCALBIN) -p path)" go test ./... -coverprofile cover.out
 
-.PHONY: test.scorecard ## Runs the operator scorecard test.
-test.scorecard: operator-sdk
+.PHONY: test.scorecard
+test.scorecard: operator-sdk ## Run the operator scorecard test against an existing cluster.
```
Comment (jwendell) [inline Makefile.core.mk]: Keep the original help. We no longer need an existing cluster. The test deploys a kind cluster.

File: Makefile.core.mk [hunk 3]
```diff
@@ -148,16 +148,16 @@ export
 test: envtest ## Run tests.
```
Comment (jwendell) [inline Makefile.core.mk]: Are these `unit` tests?

File: Makefile.core.mk [hunk 4]
```diff
@@ -242,57 +242,57 @@ ifndef ignore-not-found
...
 
 .PHONY: deploy-example-openshift
-deploy-example-openshift: ## Deploy an example Istio resource on OpenShift
+deploy-example-openshift: ## Deploy an example Istio resource to an existing OCP cluster.
 	kubectl create ns istio-system || echo "namespace istio-system already exists"
 	kubectl apply -n istio-system -f chart/samples/istio-sample-openshift.yaml
 
 .PHONY: deploy-example-kubernetes
-deploy-example-kubernetes: ## Deploy an example Istio resource on Kubernetes
+deploy-example-kubernetes: ## Deploy an example Istio resource on an ecisting cluster.
```
Comment (jwendell) [inline Makefile.core.mk]: typo

File: Makefile.core.mk [hunk 5]
```diff
@@ -341,11 +341,11 @@ gen-charts: ## Pull charts from maistra/istio repository
 	@# calls copy-crds.sh with the version specified in the .crdSourceVersion field in versions.yaml
 	@hack/copy-crds.sh "resources/$$(yq eval '.crdSourceVersion' versions.yaml)/charts"
 
-.PHONY: gen ## Generate everything
-gen: controller-gen gen-api gen-charts gen-manifests gen-code bundle
+.PHONY: gen
+gen: controller-gen gen-api gen-charts gen-manifests gen-code bundle ## Generate everything.
```
Comment (jwendell) [inline Makefile.core.mk:343]: That reminds me AI :D

**Comment by jwendell** [comment]
I did not get this pretty output. I got:

```
$ BUILD_WITH_CONTAINER=0 make help
common/Makefile.common.mk:120: warning: overriding recipe for target 'help'
Makefile.overrides.mk:19: warning: ignoring old recipe for target 'help'
egrep: warning: egrep is obsolescent; using grep -E
build                Build manager binary.
build                Build manager binary.
bundle-build         Build the bundle image.
bundle-build         Build the bundle image.
bundle               Generate bundle manifests and metadata, then validate generated files.
...
```

**Comment by dgn** [comment]
> I did not get this pretty output. I got:
> 
> ```
> $ BUILD_WITH_CONTAINER=0 make help
> common/Makefile.common.mk:120: warning: overriding recipe for target 'help'
> Makefile.overrides.mk:19: warning: ignoring old recipe for target 'help'
> egrep: warning: egrep is obsolescent; using grep -E
> build                Build manager binary.
> build                Build manager binary.
> bundle-build         Build the bundle image.
> bundle-build         Build the bundle image.
> bundle               Generate bundle manifests and metadata, then validate generated files.
> ...
> ```

just leave out the `BUILD_WITH_CONTAINER=0`

**Comment by dgn** [comment]
Try again now! It should work even with BUILD_WITH_CONTAINER=0.

I had to make our imports from common explicit instead of using `include`, that also avoids the `egrep` error. Tbh I'm much more comfortable with that anyway, the forced import of everything is a bit limiting

File: Makefile.core.mk [hunk 6]
```diff
@@ -12,8 +12,6 @@
 ## See the License for the specific language governing permissions and
 ## limitations under the License.
 
--include Makefile.overrides
```
Comment (dgn) [inline Makefile.core.mk:15]: removed this as it never did anything, it was missing the `.mk` suffix

File: Makefile.core.mk (hunk 2 above)
Comment (dgn) [reply Makefile.core.mk]: done

File: Makefile.core.mk (hunk 3 above)
Comment (dgn) [reply Makefile.core.mk]: done

File: Makefile.core.mk (hunk 4 above)
Comment (dgn) [reply Makefile.core.mk]: fixed

File: Makefile.core.mk [hunk 7]
```diff
@@ -503,19 +501,22 @@ catalog-push: ## Push a catalog image.
...
-lint: lint-scripts lint-copyright-banner lint-go lint-yaml lint-helm lint-bundle lint-watches ## runs all linters
+lint: lint-scripts lint-copyright-banner lint-go lint-yaml lint-helm lint-bundle lint-watches ## Run all linters.
 
 .PHONY: format
-format: format-go tidy-go ## Auto formats all code. This should be run before sending a PR.
+format: format-go tidy-go ## Auto-format all code. This should be run before sending a PR.
 
 .SILENT: helm $(HELM) $(LOCALBIN) deploy-yaml
 
-include common/Makefile.common.mk
```
Comment (jwendell) [inline Makefile.core.mk:521]: This is not good :(

We no longer are able to invoke other targets in that Makefile, eg `lint-all` or `update-common` which is used in tools/update_deps.sh ...

Comment (dgn) [reply Makefile.core.mk:521]: added them back

**Review by jwendell — APPROVED** [review]

**Pull request merged**